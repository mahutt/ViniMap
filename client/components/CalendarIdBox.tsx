import React, { useState } from 'react';
import {
  Modal,
  View,
  Text,
  TextInput,
  TouchableOpacity,
  TouchableWithoutFeedback,
  StyleSheet,
} from 'react-native';

interface SimpleModalProps {
  visible: boolean;
  onClose: () => void;
  onSave: (value: string) => void; // ✅ Callback to pass data back
}

const SimpleModal: React.FC<SimpleModalProps> = ({ visible, onClose, onSave }) => {
  const [localInput, setLocalInput] = useState(''); // ✅ Local state for input

  const handleSave = () => {
    onSave(localInput); // ✅ Send input back to parent
    onClose(); // ✅ Close modal after saving
  };

  return (
    <Modal transparent={true} visible={visible} animationType="slide">
      <TouchableWithoutFeedback onPress={onClose}>
        <View style={styles.modalBackground}>
          <TouchableWithoutFeedback onPress={() => {}}>
            <View style={styles.modalContent}>
              <Text style={styles.modalTitle}>Enter Something</Text>
              <TextInput
                style={styles.input}
                placeholder="Type here..."
                value={localInput}
                onChangeText={setLocalInput}
              />
              <View style={styles.buttonContainer}>
                <TouchableOpacity onPress={onClose} style={styles.cancelButton}>
                  <Text style={styles.buttonText}>Cancel</Text>
                </TouchableOpacity>
                <TouchableOpacity onPress={handleSave} style={styles.saveButton}>
                  <Text style={styles.buttonText}>Save</Text>
                </TouchableOpacity>
              </View>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </TouchableWithoutFeedback>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalBackground: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,0.5)',
  },
  modalContent: {
    width: '80%',
    backgroundColor: '#fff',
    padding: 20,
    borderRadius: 8,
    alignItems: 'center',
  },
  modalTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 12,
  },
  input: {
    width: '100%',
    height: 40,
    borderColor: '#ddd',
    borderWidth: 1,
    borderRadius: 6,
    paddingHorizontal: 10,
    marginBottom: 12,
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '100%',
  },
  cancelButton: {
    backgroundColor: '#ccc',
    paddingVertical: 8,
    paddingHorizontal: 16,
    borderRadius: 6,
    marginRight: 10,
  },
  saveButton: {
    backgroundColor: '#852C3A',
    paddingVertical: 8,
    paddingHorizontal: 16,
    borderRadius: 6,
  },
  buttonText: {
    color: '#fff',
    fontWeight: 'bold',
  },
});

export default SimpleModal;
