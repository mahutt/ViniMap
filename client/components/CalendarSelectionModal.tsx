import React, { useState, useEffect } from 'react';
import {
  Modal,
  View,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  StyleSheet,
  FlatList,
  Alert,
  Image,
} from 'react-native';
import GoogleAuthService from '@/services/GoogleService';

interface CalendarItem {
  id: string;
  summary: string;
  description?: string;
  primary?: boolean;
  selected?: boolean;
}

interface CalendarSelectionModalProps {
  visible: boolean;
  onClose: () => void;
  onSelect: (calendarId: string) => void;
  onEnterCalendarId: () => void;
  onSignOut: () => void;
}

const CalendarSelectionModal: React.FC<CalendarSelectionModalProps> = ({
  visible,
  onClose,
  onSelect,
  onEnterCalendarId,
  onSignOut,
}) => {
  const [calendars, setCalendars] = useState<CalendarItem[]>([]);

  const userInfo = GoogleAuthService.getUserInfoFromStorage();
  const selectedCalendarId = GoogleAuthService.getSelectedCalendarId();

  useEffect(() => {
    if (userInfo) {
      loadCalendars();
    }
  }, [userInfo]);

  const loadCalendars = async () => {
    try {
      const calendarList = await GoogleAuthService.fetchUserCalendars();
      const calendarItems: CalendarItem[] = calendarList.map((calendar: any) => ({
        ...calendar,
        selected: calendar.id === selectedCalendarId,
      }));
      setCalendars(calendarItems);
    } catch (error) {
      console.error('Failed to load calendars:', error);
      Alert.alert('Error', 'Failed to load your calendars. Please try again.');
    }
  };

  const handleSelectCalendar = (calendarId: string) => {
    GoogleAuthService.saveSelectedCalendarId(calendarId);
    onSelect(calendarId);
    onClose();
  };

  const handleSignOut = () => {
    onSignOut();
    onClose();
  };

  return (
    <Modal transparent={true} visible={visible}>
      <TouchableWithoutFeedback onPress={onClose}>
        <View style={styles.modalBackground}>
          <TouchableWithoutFeedback onPress={() => {}}>
            <View style={styles.modalContent}>
              {userInfo && (
                <View style={styles.userInfoContainer}>
                  {userInfo.picture ? (
                    <Image source={{ uri: userInfo.picture }} style={styles.profileImage} />
                  ) : (
                    <View style={[styles.profileFallback, { backgroundColor: '#6A8CAF' }]}>
                      <Text style={styles.profileInitial}>
                        {userInfo.name ? userInfo.name.charAt(0).toUpperCase() : 'U'}
                      </Text>
                    </View>
                  )}
                  <Text style={styles.userName}>{userInfo.name}</Text>
                  <Text style={styles.userEmail}>{userInfo.email}</Text>
                </View>
              )}

              <Text style={styles.sectionTitle}>Your Calendars</Text>

              <>
                {calendars.length === 0 ? (
                  <Text style={styles.infoText}>No calendars found</Text>
                ) : (
                  <FlatList
                    data={calendars}
                    keyExtractor={(item) => item.id}
                    style={styles.list}
                    renderItem={({ item }) => (
                      <TouchableOpacity
                        style={styles.calendarItem}
                        onPress={() => handleSelectCalendar(item.id)}>
                        <View style={styles.calendarItemContent}>
                          <View style={styles.checkboxContainer}>
                            {item.selected && (
                              <View style={styles.checkbox}>
                                <Text style={styles.checkmark}>âœ“</Text>
                              </View>
                            )}
                          </View>
                          <View style={styles.calendarTextContainer}>
                            <Text style={styles.calendarName}>
                              {item.summary} {item.primary ? '(Primary)' : ''}
                            </Text>
                            {item.description && (
                              <Text style={styles.calendarDescription}>{item.description}</Text>
                            )}
                          </View>
                        </View>
                      </TouchableOpacity>
                    )}
                  />
                )}
              </>

              <View style={styles.footerButtons}>
                <TouchableOpacity style={styles.footerButton} onPress={onEnterCalendarId}>
                  <Text style={styles.footerButtonText}>Enter Calendar ID Manually</Text>
                </TouchableOpacity>

                <TouchableOpacity style={styles.signOutButton} onPress={handleSignOut}>
                  <Text style={styles.signOutText}>Sign Out</Text>
                </TouchableOpacity>
              </View>
            </View>
          </TouchableWithoutFeedback>
        </View>
      </TouchableWithoutFeedback>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalBackground: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,0.5)',
  },
  modalContent: {
    width: '90%',
    maxHeight: '80%',
    backgroundColor: '#fff',
    padding: 0,
    borderRadius: 12,
    overflow: 'hidden',
  },
  userInfoContainer: {
    backgroundColor: '#f5f5f5',
    padding: 20,
    alignItems: 'center',
    borderBottomWidth: 1,
    borderBottomColor: '#ddd',
  },
  profileImage: {
    width: 60,
    height: 60,
    borderRadius: 30,
    marginBottom: 10,
  },
  profileFallback: {
    width: 60,
    height: 60,
    borderRadius: 30,
    marginBottom: 10,
    justifyContent: 'center',
    alignItems: 'center',
  },
  profileInitial: {
    color: 'white',
    fontSize: 26,
    fontWeight: 'bold',
  },
  userName: {
    fontSize: 18,
    fontWeight: '600',
    marginBottom: 4,
  },
  userEmail: {
    fontSize: 14,
    color: '#666',
  },
  sectionTitle: {
    fontSize: 16,
    fontWeight: '600',
    padding: 16,
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
  infoText: {
    textAlign: 'center',
    marginVertical: 20,
    color: '#666',
  },
  loader: {
    marginVertical: 20,
  },
  list: {
    maxHeight: 300,
  },
  calendarItem: {
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
  calendarItemContent: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 12,
  },
  checkboxContainer: {
    width: 24,
    height: 24,
    marginRight: 12,
    justifyContent: 'center',
    alignItems: 'center',
  },
  checkbox: {
    width: 20,
    height: 20,
    borderRadius: 10,
    backgroundColor: '#852C3A',
    justifyContent: 'center',
    alignItems: 'center',
  },
  checkmark: {
    color: 'white',
    fontSize: 12,
    fontWeight: 'bold',
  },
  calendarTextContainer: {
    flex: 1,
  },
  calendarName: {
    fontSize: 16,
    fontWeight: '500',
  },
  calendarDescription: {
    fontSize: 12,
    color: '#666',
    marginTop: 4,
  },
  footerButtons: {
    borderTopWidth: 1,
    borderTopColor: '#eee',
    padding: 16,
  },
  footerButton: {
    paddingVertical: 12,
    alignItems: 'center',
    borderBottomWidth: 1,
    borderBottomColor: '#eee',
  },
  footerButtonText: {
    color: '#852C3A',
    fontWeight: '500',
  },
  signOutButton: {
    paddingVertical: 12,
    alignItems: 'center',
    marginTop: 8,
  },
  signOutText: {
    color: '#FF3B30',
    fontWeight: '500',
  },
});

export default CalendarSelectionModal;
