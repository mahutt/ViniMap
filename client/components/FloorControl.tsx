import { useMap } from '@/modules/map/MapContext';
import { Text, View, StyleSheet, TouchableOpacity } from 'react-native';
import { IconSymbol } from '@/components/ui/IconSymbol';
import React from 'react';

export default function FloorControl() {
  const { level, setLevel, indoorMap } = useMap();

  if (indoorMap === null || level === null) {
    return null;
  }

  const handleLevelChange = (level: number) => {
    if (level < indoorMap.levelsRange.min || level > indoorMap.levelsRange.max) {
      return;
    }
    setLevel(level);
  };

  const atMaxLevel = level === indoorMap.levelsRange.max;
  const atMinLevel = level === indoorMap.levelsRange.min;

  return (
    <View style={styles.controlContainer}>
      <TouchableOpacity
        disabled={atMaxLevel}
        style={{ ...styles.button, paddingTop: 10, opacity: atMaxLevel ? 0.3 : 1 }}
        onPress={() => handleLevelChange(level + 1)}>
        <IconSymbol name="chevron.up" size={20} color="#912338" />
      </TouchableOpacity>
      <View style={styles.levelContainer}>
        <Text style={styles.levelText}>{level}</Text>
      </View>
      <TouchableOpacity
        disabled={atMinLevel}
        style={{ ...styles.button, paddingBottom: 10, opacity: atMinLevel ? 0.3 : 1 }}
        onPress={() => handleLevelChange(level - 1)}>
        <IconSymbol name="chevron.down" size={20} color="#912338" />
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  controlContainer: {
    position: 'absolute',
    top: '25%',
    right: 36,

    backgroundColor: 'rgba(255, 255, 255, 0.8)',
    borderRadius: 20,
    borderWidth: 2,
    borderColor: '#912338',
    zIndex: 10,
  },
  button: {
    width: 50,
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
  },
  levelContainer: {
    width: 50,
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
  },
  levelText: {
    color: '#912338',
    fontSize: 16,
    fontWeight: 'bold',
  },
});
