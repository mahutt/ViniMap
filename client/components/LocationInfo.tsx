import { MapState, useMap } from '@/modules/map/MapContext';
import { Ionicons } from '@expo/vector-icons';
import React from 'react';
import { Text, StyleSheet, View, TouchableOpacity } from 'react-native';
import { IconSymbol } from './ui/IconSymbol';

const iconMap: Record<string, keyof typeof Ionicons.glyphMap> = {
  bicycle_rental: 'bicycle-outline',
  subway_station: 'subway-outline',
  bus_station: 'bus-outline',
  restaurant: 'restaurant-outline',
  park: 'leaf-outline',
  library: 'book-outline',
  location: 'location-outline',
  university: 'school-outline',
};

export function LocationInfo() {
  const { setState, endLocation } = useMap();

  function getDirections() {
    setState(MapState.RoutePlanning);
  }

  const isCustomPOI = endLocation?.data?.hours !== undefined;
  const iconName = iconMap[endLocation?.data?.type] || 'location';

  return (
    <View style={styles.container}>
      <View style={styles.headerContainer}>
        <Text style={styles.nameText}>{endLocation?.name}</Text>
        <TouchableOpacity
          style={styles.closeButton}
          onPress={() => setState(MapState.Idle)}
          testID="close-button">
          <Text style={styles.closeText}>
            <IconSymbol name="xmark" size={15} color="#333" />
          </Text>
        </TouchableOpacity>
      </View>

      <View style={{ flexDirection: 'row', alignItems: 'center', gap: 8, marginBottom: 8 }}>
        {isCustomPOI && <Ionicons name={iconName} size={16} color="#852C3A" />}
        {isCustomPOI && endLocation?.data?.description && (
          <Text style={styles.descriptionText}>{endLocation.data.description}</Text>
        )}
      </View>

      <View style={{ flexDirection: 'row', alignItems: 'center', gap: 8 }}>
        <Text style={[styles.isOpen, { color: endLocation?.data?.isOpen ? 'green' : 'red' }]}>
          {endLocation?.data?.isOpen ? 'Open Now' : 'Closed Now'}
        </Text>

        {isCustomPOI && endLocation?.data?.hours && (
          <>
            <Text>â€¢</Text>
            <Text numberOfLines={1} ellipsizeMode="tail" style={styles.hoursText}>
              Hours: {endLocation.data.hours}
            </Text>
          </>
        )}
      </View>

      <Text style={styles.addressText}>{endLocation?.data?.address}</Text>

      <TouchableOpacity style={styles.button} onPress={getDirections}>
        <Ionicons name="arrow-forward-outline" size={24} color="#fff" style={styles.icon} />
        <Text style={styles.text}>Directions</Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    display: 'flex',
    flexDirection: 'column',
    backgroundColor: 'white',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    gap: 4,
    padding: 20,
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
    zIndex: 10,
  },
  headerContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    gap: 8,
  },
  nameText: {
    flex: 1,
    fontSize: 20,
    fontWeight: 'bold',
  },
  closeButton: {
    backgroundColor: '#ccc',
    borderRadius: 15,
    width: 30,
    height: 30,
    justifyContent: 'center',
    alignItems: 'center',
  },
  closeText: {
    color: 'black',
    fontSize: 20,
    lineHeight: 20,
  },
  addressText: {
    fontSize: 16,
    color: 'gray',
  },
  isOpen: {
    fontSize: 16,
  },
  button: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#852C3A',
    paddingVertical: '3%',
    paddingHorizontal: '5%',
    borderRadius: 25,
    justifyContent: 'center',
    marginTop: '8%',
  },
  icon: {
    marginRight: 10,
  },
  text: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
  hoursText: {
    flex: 1,
    fontSize: 16,
    color: '#666',
  },
  descriptionText: {
    fontSize: 16,
    color: '#333',
  },
});
