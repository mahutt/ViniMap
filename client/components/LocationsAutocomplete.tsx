import { useEffect, useState } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { Location } from '@/modules/map/Types';
import { getLocations } from '@/modules/map/MapService';
import LocalLocations from '@/services/LocalLocations';

export default function LocationsAutocomplete({
  query,
  callback,
}: Readonly<{
  query: string;
  callback: (location: Location) => void;
}>) {
  const [locations, setLocations] = useState<Location[]>([]);
  useEffect(() => {
    if (query) {
      const localLocations = LocalLocations.getInstance().autocomplete(query);
      if (localLocations.length >= 5) {
        setLocations(localLocations.slice(0, 5));
      } else {
        getLocations(query).then((locations) => {
          setLocations(localLocations.concat(locations).slice(0, 5));
        });
      }
    } else {
      setLocations([]);
    }
  }, [query]);

  if (locations.length === 0) {
    return null;
  }

  return (
    <View style={styles.locationsContainer}>
      {locations.map((location, index) => (
        <View
          key={`location-${index}`}
          style={styles.locationItem}
          onTouchEnd={() => {
            callback(location);
            setLocations([]);
          }}>
          <Text>{location.name}</Text>
        </View>
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  locationsContainer: {
    position: 'absolute',
    top: '100%',
    left: 0,
    right: 0,
    margin: 0,
    paddingVertical: 10,
    backgroundColor: 'white',
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 12,
    zIndex: 2,
    overflow: 'hidden',
  },
  locationItem: {
    paddingHorizontal: 16,
    paddingVertical: 10,
    margin: 0,
  },
});
