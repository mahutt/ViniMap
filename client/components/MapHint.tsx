import { View, Text, StyleSheet } from 'react-native';
import { useMap, MapState } from '@/modules/map/MapContext';

function getHint(state: MapState): string | null {
  switch (state) {
    case MapState.SelectingStartLocation:
      return 'Tap anywhere to set your start location';
    case MapState.SelectingEndLocation:
      return 'Tap anywhere to set your destination';
    default:
      return null;
  }
}

export default function MapHint() {
  const { state } = useMap();
  const hint = getHint(state);
  if (!hint) return null;
  return (
    <View style={styles.hintContainer}>
      <Text style={styles.hintText}>{hint}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  hintContainer: {
    position: 'absolute',
    top: 140,
    left: 20,
    right: 20,
    backgroundColor: 'rgba(0, 0, 0, 0.7)',
    padding: 12,
    borderRadius: 8,
    zIndex: 1000,
  },
  hintText: {
    color: 'white',
    fontSize: 16,
    textAlign: 'center',
  },
});
