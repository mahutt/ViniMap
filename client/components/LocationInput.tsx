import { MapState, useMap } from '@/modules/map/MapContext';
import { Location } from '@/modules/map/Types';
import { Ionicons } from '@expo/vector-icons';
import React, { useEffect } from 'react';
import { TextInput, View, StyleSheet, TouchableOpacity, Text } from 'react-native';
import LocationsAutocomplete from './LocationsAutocomplete';

const CURRENT_LOCATION_NAME = 'Current location';

export default function LocationInput({
  location,
  setLocation,
  ionIconName,
  placeholder,
  isStartLocation = false,
}: Readonly<{
  location: Location | null;
  setLocation: (location: Location) => void;
  ionIconName: 'pin-outline' | 'pin';
  placeholder: string;
  isStartLocation?: boolean;
}>) {
  const inputRef = React.useRef<TextInput>(null);
  const [query, setQuery] = React.useState<string>('');
  const [isFocused, setIsFocused] = React.useState(false);
  const { setState, userLocation } = useMap();

  useEffect(() => {
    if (location) {
      setQuery(location.name ?? '');
    }
  }, [location]);

  const handleFocus = () => {
    setIsFocused(true);

    setTimeout(() => {
      if (inputRef.current?.setSelection) {
        inputRef.current.setSelection(0, query.length);
      }
    }, 100);
  };

  const handleBlur = () => {
    setIsFocused(false);
    if (location) {
      setQuery(location.name ?? '');
    }
  };

  const handleCurrentLocation = async () => {
    if (isStartLocation) {
      if (!userLocation) {
        console.error('User location not available');
        return;
      }
      setLocation(userLocation);
      inputRef.current?.blur();
    }
  };

  // helper functions
  const shouldShowMapPrompt = () => {
    if (!isFocused) return false;
    if (query === '') return true;
    if (query === location?.name) return true;
    return false;
  };

  const shouldShowAutocomplete = () => {
    if (!isFocused) return false;
    if (query !== '' && query !== location?.name) return true;
    return false;
  };

  return (
    <View style={styles.locationInputContainer}>
      <Ionicons name={ionIconName} size={16} color="#666" />
      <TextInput
        ref={inputRef}
        style={styles.input}
        placeholder={placeholder}
        placeholderTextColor="#666"
        value={query}
        autoCorrect={false}
        onChangeText={setQuery}
        onFocus={handleFocus}
        onBlur={handleBlur}
      />

      {shouldShowMapPrompt() && (
        <View style={styles.optionsContainer}>
          <TouchableOpacity
            style={styles.optionItem}
            onPress={() => {
              setState(
                isStartLocation ? MapState.SelectingStartLocation : MapState.SelectingEndLocation
              );
            }}>
            <Ionicons name="map-outline" size={20} color="#666" />
            <Text style={styles.optionText}>Choose on map</Text>
          </TouchableOpacity>
          {isStartLocation && location?.name !== CURRENT_LOCATION_NAME && (
            <TouchableOpacity style={styles.optionItem} onPress={handleCurrentLocation}>
              <Ionicons name="locate-outline" size={20} color="#666" />
              <Text style={styles.optionText}>Use Current Location</Text>
            </TouchableOpacity>
          )}
        </View>
      )}

      {shouldShowAutocomplete() && (
        <LocationsAutocomplete
          query={query}
          callback={(selectedLocation) => {
            setLocation(selectedLocation);
            setQuery(selectedLocation.name ?? '');
            setTimeout(() => {
              inputRef.current?.blur();
            }, 0);
          }}
        />
      )}
    </View>
  );
}
const styles = StyleSheet.create({
  locationInputContainer: {
    position: 'relative',
    flex: 1,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: 10,
    paddingHorizontal: 10,
    borderColor: '#999',
    borderWidth: 1,
    borderRadius: 12,
  },
  input: {
    flex: 1,
    paddingVertical: 10,
    fontSize: 16,
  },
  optionsContainer: {
    position: 'absolute',
    top: '100%',
    left: 0,
    right: 0,
    backgroundColor: 'white',
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 12,
    zIndex: 2,
    marginTop: 4,
  },
  optionItem: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 10,
    padding: 16,
  },
  optionText: {
    fontSize: 16,
    color: '#666',
  },
});
