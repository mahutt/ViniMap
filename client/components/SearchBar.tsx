import React from 'react';
import { View, TextInput, StyleSheet } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import LocationsAutocomplete from './LocationsAutocomplete';
import { useMap, MapState } from '@/modules/map/MapContext';

export function SearchBar() {
  const { endLocation, setEndLocation, setState, flyTo } = useMap();
  const [query, setQuery] = React.useState<string>('');
  const textInputRef = React.useRef<TextInput>(null);

  return (
    <View style={styles.searchContainer}>
      <View style={styles.searchBar}>
        <Ionicons name="search" size={20} color="#666" />
        <TextInput
          ref={textInputRef}
          style={styles.input}
          placeholder="Search here"
          placeholderTextColor="#666"
          value={query}
          onChangeText={(newQuery) => setQuery(newQuery)}
          autoCorrect={false}
        />
        {query !== '' && query !== endLocation?.name && (
          <LocationsAutocomplete
            query={query}
            callback={(location) => {
              setQuery(location.name ?? '');
              setEndLocation(location);
              // setState(MapState.Information);
              flyTo(location.coordinates, 17);
              setTimeout(() => {
                textInputRef.current?.blur();
              }, 0);
            }}
          />
        )}
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  searchContainer: {
    position: 'absolute',
    top: 76,
    left: 36,
    right: 36,
    zIndex: 3,
  },
  searchBar: {
    position: 'relative',
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: 'white',
    borderRadius: 25,
    paddingHorizontal: 15,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  input: {
    flex: 1,
    marginLeft: 10,
    fontSize: 16,
    paddingVertical: 10,
  },
});
