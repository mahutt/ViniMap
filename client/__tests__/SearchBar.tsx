import { render } from '@testing-library/react-native';
import { SearchBar } from '@/components/SearchBar';

jest.mock('@expo/vector-icons', () => ({
  Ionicons: 'Ionicons',
}));

describe('<SearchBar />', () => {
  test('SearchBar renders correctly', () => {
    const tree = render(<SearchBar onSearch={(query) => {}} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
