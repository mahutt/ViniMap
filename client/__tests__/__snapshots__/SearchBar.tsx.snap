// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> SearchBar renders correctly 1`] = `
<View
  style={
    {
      "left": 20,
      "position": "absolute",
      "right": 20,
      "top": 60,
      "zIndex": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 25,
        "elevation": 5,
        "flexDirection": "row",
        "paddingHorizontal": 15,
        "paddingVertical": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <Ionicons
      color="#666"
      name="search"
      size={20}
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Search here"
      placeholderTextColor="#666"
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "marginLeft": 10,
        }
      }
    />
  </View>
</View>
`;
